unhandled exception error for sequence construction let binding

The following code generates the fatal unhandled exception:

```dafny
function method DecodeRecursively(s: seq<int>): (b: seq<int>) {
    seq(|s|, i requires 0 <= i < |s| =>
        var d := s[0]; // this line produces the error below
        s[i]
    )
}
```
```
[ERROR] FATAL UNHANDLED EXCEPTION: System.AggregateException: One or more errors occurred. (Assertion failed.) (Assertion failed.) ---> System.Diagnostics.Contracts.ContractException: Assertion failed.
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailureImplementation (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0002f] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailure (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x00000] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Diagnostics.Contracts.Contract.ReportFailure (System.Diagnostics.Contracts.ContractFailureKind failureKind, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0003a] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Diagnostics.Contracts.Contract.Assert (System.Boolean condition) [0x00003] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at Microsoft.Boogie.VCExprAST.VCExprLetBinding..ctor (Microsoft.Boogie.VCExprAST.VCExprVar v, Microsoft.Boogie.VCExprAST.VCExpr e) [0x00027] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitLetExpr (Microsoft.Boogie.LetExpr node) [0x000de] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.LetExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateNAryExpr (Microsoft.Boogie.NAryExpr node) [0x001b6] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitNAryExpr (Microsoft.Boogie.NAryExpr node) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.NAryExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateNAryExpr (Microsoft.Boogie.NAryExpr node) [0x001b6] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitNAryExpr (Microsoft.Boogie.NAryExpr node) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.NAryExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateQuantifierExpr (Microsoft.Boogie.QuantifierExpr node) [0x000b2] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitQuantifierExpr (Microsoft.Boogie.QuantifierExpr node) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitForallExpr (Microsoft.Boogie.ForallExpr node) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.ForallExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.DeclFreeProverContext.AddAxiom (Microsoft.Boogie.Axiom ax, System.String attributes) [0x0001c] in <f0c36bd8c8594201a0cc3c571868072c>:0
  at Microsoft.Boogie.Checker.Setup (Microsoft.Boogie.Program prog, Microsoft.Boogie.ProverContext ctx) [0x000d5] in <f0c36bd8c8594201a0cc3c571868072c>:0
  at Microsoft.Boogie.Checker..ctor (VC.ConditionGeneration vcgen, Microsoft.Boogie.Program prog, System.String logFilePath, System.Boolean appendLogFile, System.Int32 timeout, System.Int32 rlimit, Microsoft.Boogie.ProverContext ctx) [0x0013d] in <f0c36bd8c8594201a0cc3c571868072c>:0
  at VC.ConditionGeneration.FindCheckerFor (System.Int32 timeout, System.Int32 rlimit, System.Boolean isBlocking, System.Int32 waitTimeinMs, System.Int32 maxRetries) [0x001aa] in <f0c36bd8c8594201a0cc3c571868072c>:0
  at VC.VCGen.VerifyImplementation (Microsoft.Boogie.Implementation impl, Microsoft.Boogie.VerifierCallback callback) [0x00610] in <f0c36bd8c8594201a0cc3c571868072c>:0
  at VC.ConditionGeneration.VerifyImplementation (Microsoft.Boogie.Implementation impl, System.Collections.Generic.List`1[Microsoft.Boogie.Counterexample]& errors, System.String requestId) [0x00019] in <f0c36bd8c8594201a0cc3c571868072c>:0
  at Microsoft.Boogie.ExecutionEngine.VerifyImplementation (Microsoft.Boogie.Program program, Microsoft.Boogie.PipelineStatistics stats, Microsoft.Boogie.ErrorReporterDelegate er, System.String requestId, System.Collections.Generic.Dictionary`2[TKey,TValue] extractLoopMappingInfo, Microsoft.Boogie.Implementation[] stablePrioritizedImpls, System.Int32 index, Microsoft.Boogie.ExecutionEngine+OutputCollector outputCollector, System.Collections.Generic.List`1[T] checkers, System.String programId) [0x00243] in <95704e96274a4f489f709d7a975c8130>:0
  at Microsoft.Boogie.ExecutionEngine+<>c__DisplayClass27_2.<InferAndVerify>b__5 (System.Object dummy) [0x00056] in <95704e96274a4f489f709d7a975c8130>:0
  at System.Threading.Tasks.Task.InnerInvoke () [0x00025] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Threading.Tasks.Task.Execute () [0x00000] in <f7cedb26ce694cd281e4450870c6fe49>:0
   --- End of inner exception stack trace ---
  at System.AggregateException.Handle (System.Func`2[T,TResult] predicate) [0x00064] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at Microsoft.Boogie.ExecutionEngine.InferAndVerify (Microsoft.Boogie.Program program, Microsoft.Boogie.PipelineStatistics stats, System.String programId, Microsoft.Boogie.ErrorReporterDelegate er, System.String requestId) [0x00523] in <95704e96274a4f489f709d7a975c8130>:0
  at Microsoft.Dafny.DafnyDriver.BoogiePipelineWithRerun (Microsoft.Boogie.Program program, System.String bplFileName, Microsoft.Boogie.PipelineStatistics& stats, System.String programId) [0x000bf] in <2749e30f97454112a0477fdcfcedd3ab>:0
  at Microsoft.Dafny.DafnyDriver.BoogieOnce (System.String baseFile, System.String moduleName, Microsoft.Boogie.Program boogieProgram, System.String programId, Microsoft.Boogie.PipelineStatistics& stats, Microsoft.Boogie.PipelineOutcome& oc) [0x00083] in <2749e30f97454112a0477fdcfcedd3ab>:0
  at Microsoft.Dafny.DafnyDriver.Boogie (System.String baseName, System.Collections.Generic.IEnumerable`1[T] boogiePrograms, System.String programId, System.Collections.Generic.Dictionary`2[System.String,Microsoft.Boogie.PipelineStatistics]& statss, Microsoft.Boogie.PipelineOutcome& oc) [0x0006e] in <2749e30f97454112a0477fdcfcedd3ab>:0
  at Microsoft.Dafny.DafnyDriver.ProcessFiles (System.Collections.Generic.IList`1[T] dafnyFiles, System.Collections.ObjectModel.ReadOnlyCollection`1[T] otherFileNames, Microsoft.Dafny.ErrorReporter reporter, System.Boolean lookForSnapshots, System.String programId) [0x00222] in <2749e30f97454112a0477fdcfcedd3ab>:0
  at Microsoft.Dafny.DafnyDriver.ThreadMain (System.String[] args) [0x0003a] in <2749e30f97454112a0477fdcfcedd3ab>:0
  at Microsoft.Dafny.DafnyDriver+<>c__DisplayClass1_0.<Main>b__0 () [0x00001] in <2749e30f97454112a0477fdcfcedd3ab>:0
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00014] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Threading.ThreadHelper.ThreadStart () [0x00008] in <f7cedb26ce694cd281e4450870c6fe49>:0
---> (Inner Exception #0) System.Diagnostics.Contracts.ContractException: Assertion failed.
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailureImplementation (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0002f] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailure (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x00000] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Diagnostics.Contracts.Contract.ReportFailure (System.Diagnostics.Contracts.ContractFailureKind failureKind, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0003a] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Diagnostics.Contracts.Contract.Assert (System.Boolean condition) [0x00003] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at Microsoft.Boogie.VCExprAST.VCExprLetBinding..ctor (Microsoft.Boogie.VCExprAST.VCExprVar v, Microsoft.Boogie.VCExprAST.VCExpr e) [0x00027] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitLetExpr (Microsoft.Boogie.LetExpr node) [0x000de] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.LetExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateNAryExpr (Microsoft.Boogie.NAryExpr node) [0x001b6] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitNAryExpr (Microsoft.Boogie.NAryExpr node) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.NAryExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateNAryExpr (Microsoft.Boogie.NAryExpr node) [0x001b6] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitNAryExpr (Microsoft.Boogie.NAryExpr node) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.NAryExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.TranslateQuantifierExpr (Microsoft.Boogie.QuantifierExpr node) [0x000b2] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitQuantifierExpr (Microsoft.Boogie.QuantifierExpr node) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.VisitForallExpr (Microsoft.Boogie.ForallExpr node) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.ForallExpr.StdDispatch (Microsoft.Boogie.StandardVisitor visitor) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.ReadOnlyVisitor.Visit (Microsoft.Boogie.Absy node) [0x00001] in <44d09d447683469684e15c5d53187581>:0
  at Microsoft.Boogie.VCExprAST.Boogie2VCExprTranslator.Translate (Microsoft.Boogie.Expr expr) [0x00001] in <bdc543d85f4141079673ec18ef7f4dde>:0
  at Microsoft.Boogie.DeclFreeProverContext.AddAxiom (Microsoft.Boogie.Axiom ax, System.String attributes) [0x0001c] in <f0c36bd8c8594201a0cc3c571868072c>:0
  at Microsoft.Boogie.Checker.Setup (Microsoft.Boogie.Program prog, Microsoft.Boogie.ProverContext ctx) [0x000d5] in <f0c36bd8c8594201a0cc3c571868072c>:0
  at Microsoft.Boogie.Checker..ctor (VC.ConditionGeneration vcgen, Microsoft.Boogie.Program prog, System.String logFilePath, System.Boolean appendLogFile, System.Int32 timeout, System.Int32 rlimit, Microsoft.Boogie.ProverContext ctx) [0x0013d] in <f0c36bd8c8594201a0cc3c571868072c>:0
  at VC.ConditionGeneration.FindCheckerFor (System.Int32 timeout, System.Int32 rlimit, System.Boolean isBlocking, System.Int32 waitTimeinMs, System.Int32 maxRetries) [0x001aa] in <f0c36bd8c8594201a0cc3c571868072c>:0
  at VC.VCGen.VerifyImplementation (Microsoft.Boogie.Implementation impl, Microsoft.Boogie.VerifierCallback callback) [0x00610] in <f0c36bd8c8594201a0cc3c571868072c>:0
  at VC.ConditionGeneration.VerifyImplementation (Microsoft.Boogie.Implementation impl, System.Collections.Generic.List`1[Microsoft.Boogie.Counterexample]& errors, System.String requestId) [0x00019] in <f0c36bd8c8594201a0cc3c571868072c>:0
  at Microsoft.Boogie.ExecutionEngine.VerifyImplementation (Microsoft.Boogie.Program program, Microsoft.Boogie.PipelineStatistics stats, Microsoft.Boogie.ErrorReporterDelegate er, System.String requestId, System.Collections.Generic.Dictionary`2[TKey,TValue] extractLoopMappingInfo, Microsoft.Boogie.Implementation[] stablePrioritizedImpls, System.Int32 index, Microsoft.Boogie.ExecutionEngine+OutputCollector outputCollector, System.Collections.Generic.List`1[T] checkers, System.String programId) [0x00243] in <95704e96274a4f489f709d7a975c8130>:0
  at Microsoft.Boogie.ExecutionEngine+<>c__DisplayClass27_2.<InferAndVerify>b__5 (System.Object dummy) [0x00056] in <95704e96274a4f489f709d7a975c8130>:0
  at System.Threading.Tasks.Task.InnerInvoke () [0x00025] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Threading.Tasks.Task.Execute () [0x00000] in <f7cedb26ce694cd281e4450870c6fe49>:0 <---
```

--- a/Test/git-issues/git-issue-351.dfy
+++ b/Test/git-issues/git-issue-351.dfy
@@ -3,8 +3,8 @@
 
 
 function method DecodeRecursively(s: seq<int>): (b: seq<int>) {
-    seq(|s|, i requires 0 <= i < |s| =>
-        var d := s[0]; // this line produces the error below
-        s[i]
-    )
+  seq(|s|, i requires 0 <= i < |s| =>
+    var d := s[0]; // this line produced the reported error, prior to the fix
+    s[i]
+  )
 }

--- a/Test/git-issues/git-issue-351.dfy
+++ b/Test/git-issues/git-issue-351.dfy
@@ -1,0 +1,10 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+
+function method DecodeRecursively(s: seq<int>): (b: seq<int>) {
+    seq(|s|, i requires 0 <= i < |s| =>
+        var d := s[0]; // this line produces the error below
+        s[i]
+    )
+}

--- a/Test/git-issues/git-issue-351.dfy.expect
+++ b/Test/git-issues/git-issue-351.dfy.expect
@@ -1,0 +1,2 @@
+
+Dafny program verifier finished with 1 verified, 0 errors

